### PR TITLE
Reject duplicate parallel gateway activate command

### LINKED ISSUE #6778: IllegalStateException: Not expected to have an active sequence flow count lower then zero!
**Describe the bug**
On reprocessing, the stream processor ran in an endless loop.

```
java.lang.IllegalStateException: Not expected to have an active sequence flow count lower then zero!
``` 

Followed by

```
Catched exception class io.zeebe.engine.processing.bpmn.BpmnProcessingException with message 
Expected to take transition to 'ELEMENT_ACTIVATED' but element instance is in state 'SEQUENCE_FLOW_TAKEN'. 
[context: {intent=SEQUENCE_FLOW_TAKEN, elementId=Activity_0v8ebms, bpmnElementType=SUB_PROCESS, 
elementInstanceKey=4503599646176568, flowScopeKey=4503599646085191, processInstanceKey=4503599646085191,
parentProcessInstanceKey=-1, parentElementInstanceKey=-1, bpmnProcessId=BENCH-PERF, processVersion=1, 
processDefinitionKey=2251799813685261}], will retry...
```

**To Reproduce**

Not clear.

image

It seems to be related to the termination of an (event-) subprocess.

```
"Don't process event because of an illegal state transition: Expected element instance to be in state 'ELEMENT_TERMINATED' or one of '[]' but was 'ELEMENT_TERMINATING'. [context: {intent=ELEMENT_TERMINATED, elementId=Activity_0sul25s, bpmnElementType=SUB_PROCESS, elementInstanceKey=4503599646086641, flowScopeKey=4503599646085191, processInstanceKey=4503599646085191, parentProcessInstanceKey=-1, parentElementInstanceKey=-1, bpmnProcessId=BENCH-PERF, processVersion=1, processDefinitionKey=2251799813685261}]"
```

**Expected behavior**
This issue is probably caused by the [current work in the workflow engine](https://github.com/camunda-cloud/zeebe/milestone/51) that applies event sourcing to the stream processors.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.IllegalStateException: Not expected to have an active sequence flow count lower then zero!
	at io.zeebe.engine.state.instance.ElementInstance.decrementActiveSequenceFlows(ElementInstance.java:168) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.activateElementInstanceInFlowScope(BpmnStateTransitionBehavior.java:245) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.sequenceflow.SequenceFlowProcessor.onSequenceFlowTaken(SequenceFlowProcessor.java:96) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.sequenceflow.SequenceFlowProcessor.onActivating(SequenceFlowProcessor.java:52) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.sequenceflow.SequenceFlowProcessor.onActivating(SequenceFlowProcessor.java:30) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processEvent(BpmnStreamProcessor.java:165) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processRecord(BpmnStreamProcessor.java:106) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:54) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.ReProcessingStateMachine.reprocessRecord(ReProcessingStateMachine.java:405) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.ReProcessingStateMachine.lambda$chooseOperationForEvent$7(ReProcessingStateMachine.java:371) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[zeebe-db-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.ReProcessingStateMachine.lambda$processUntilDone$4(ReProcessingStateMachine.java:343) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:36) ~[zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.retry.EndlessRetryStrategy.run(EndlessRetryStrategy.java:50) ~[zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:94) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:191) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
```

</p>
</details>

<details><summary>Follow-up Exception</summary>
 <p>

```
io.zeebe.engine.processing.bpmn.BpmnProcessingException: Expected to take transition to 'ELEMENT_ACTIVATED' but element instance is in state 'SEQUENCE_FLOW_TAKEN'. [context: {intent=SEQUENCE_FLOW_TAKEN, elementId=Activity_0v8ebms, bpmnElementType=SUB_PROCESS, elementInstanceKey=4503599646176568, flowScopeKey=4503599646085191, processInstanceKey=4503599646085191, parentProcessInstanceKey=-1, parentElementInstanceKey=-1, bpmnProcessId=BENCH-PERF, processVersion=1, processDefinitionKey=2251799813685261}]
	at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.verifyTransition(BpmnStateTransitionBehavior.java:175) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.transitionTo(BpmnStateTransitionBehavior.java:161) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.transitionToActivated(BpmnStateTransitionBehavior.java:92) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.container.SubProcessProcessor.lambda$onActivating$1(SubProcessProcessor.java:53) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.Either$Right.ifRightOrLeft(Either.java:238) ~[zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.container.SubProcessProcessor.onActivating(SubProcessProcessor.java:52) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.container.SubProcessProcessor.onActivating(SubProcessProcessor.java:23) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processEvent(BpmnStreamProcessor.java:165) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processRecord(BpmnStreamProcessor.java:106) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:54) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.ReProcessingStateMachine.reprocessRecord(ReProcessingStateMachine.java:405) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.ReProcessingStateMachine.lambda$chooseOperationForEvent$7(ReProcessingStateMachine.java:371) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[zeebe-db-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.engine.processing.streamprocessor.ReProcessingStateMachine.lambda$processUntilDone$4(ReProcessingStateMachine.java:343) ~[zeebe-workflow-engine-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:36) ~[zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.retry.EndlessRetryStrategy.run(EndlessRetryStrategy.java:50) ~[zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:94) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:191) [zeebe-util-1.0.0-alpha4.jar:1.0.0-alpha4]
```

</p>
</details>

Error Reporting:
* https://console.cloud.google.com/errors/CMry7a29qqy7Rw?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911
* https://console.cloud.google.com/errors/CIudxdnGhOOp7AE?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911

Log: https://console.cloud.google.com/logs/query;query=%0AlogName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22ultratest%22%0Aresource.labels.namespace_name%3D%2299936d0e-e21b-493c-8b96-6d79ac11c40d-zeebe%22%0Aresource.labels.container_name%3D%22zeebe%22;timeRange=2021-04-12T10:06:14.712Z%2F2021-04-12T11:06:14.712Z;pinnedLogId=2021-04-12T10:35:15.759906Z%2F32bhcwg42mromx;cursorTimestamp=2021-04-12T10:35:15.789511Z?project=camunda-cloud-240911 

**Environment:**
 - OS: Camunda Cloud (ultratest)
 - Zeebe Version: 1.0.0-alpha4
 - Configuration:

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

Parallel gateways get activated by taking a sequence flow and checking if the number of taken flows is greater or equal to the number of incoming sequence flows. If this is the case an activate command is sent. The number of taken sequence flows get rest upon activation of the parallel gateway.

This proves troublesome when a "bad" model causes one of the incoming sequence flows to be taken twice. This could result in the activation command being sent twice. Imagine there is a parallel gateway with 2 incoming flows. What would happen is:

1. First flow is taken
2. Second flow is taken. Incoming flows == taken flows so an activate command is sent.
3. Second flow is taken again. The first activate command has not been processed yet. The number of taken flows has not been reset. As a result incoming flows < taken flows. A second activate command is sent.

This is solved by always sending an activate command when a sequence flow is taken. Once the `BpmnStreamProcessor` tries to process the record it will check if the state is valid. Here a check has been added to verify that when we receive an activate command for a parallel gateway we will first check if all the incoming flows have been taken. If this is not the case we will reject the command.

## Related issues

closes #6778 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
@korthout I have removed you as a reviewer for now, since the solution is flawed. Sorry!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java [hunk 1]
```diff
@@ -37,15 +39,19 @@ public ProcessInstanceStateTransitionGuard(final BpmnStateBehavior stateBehavior
...
+  private Either<String, ?> checkStateTransition(
+      final BpmnElementContext context, final ExecutableFlowElement element) {
     switch (context.getIntent()) {
       case ACTIVATE_ELEMENT:
-        return hasActiveFlowScopeInstance(context);
-
+        final Either<String, ?> hasActiveFlowScopeInstance = hasActiveFlowScopeInstance(context);
+        return hasActiveFlowScopeInstance.isLeft()
+            ? hasActiveFlowScopeInstance
+            : canActivateParallelGateway(context, element);
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java:54]: 🔧 Just a quick tip: You can use `flatMap(ok -> canActivateParallelGateway())` to achieve the same as the `if(failed) ? return early : else do another check that can fail and return that`

**Comment by korthout** [comment]
@remcowesterhoud I like the idea of rejecting the activate command for the parallel gateway when it isn't yet ready to activate. It would allow us to remove the logic that determines whether or not to write the activate command, and entirely move it to the processor. 

This also makes it more in line with other concurrent behaviors (e.g. an element might be planned to be completed because a COMPLETE_ELEMENT was written for it on the log, but before it is processed a TERMINATE_ELEMENT is processed due to some interruption like a boundary event, we then reject the COMPLETE_ELEMENT command when we try to process it later). 

To make this idea work for Start Process Instance Anywhere, we need something special. We'll need to increment the number of taken sequence flows before processing the ACTIVATE_ELEMENT of the parallel gateway:
- for each parallel gateway that is targetted by a start instruction,
- for each incoming sequence flow,
- within the related flow scope instance.

We could do that by introducing a new event applier for ProcessInstanceCreation CREATED events, which is written (and applied) before the ACTIVATE_ELEMENT of the parallel gateway is processed.

The hard part is to find the related flow scope instance, because you don't have access to the key of the parallel gateway's element instance (i.e. the parallel gateway isn't created yet). There is no direct access to this, but you could traverse the process instance's process model using `ElementInstanceState.getChildren(key)`.

@remcowesterhoud If you want we can talk about it in person.

**Comment by remcowesterhoud** [comment]
@korthout Thanks for your help ❤️ 
 
I have resolved the issues, please have a look!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableProcess.java [hunk 2]
```diff
@@ -33,6 +33,10 @@ public AbstractFlowElement getElementById(final DirectBuffer id) {
     return flowElements.get(id);
   }
 
+  public AbstractFlowElement getElementById(final String id) {
+    return flowElements.get(wrapString(id));
+  }
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableProcess.java:38]: 👍 That's a great idea!

File: engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  private ElementInstance findParentFlowScope(
+      final ElementInstance processInstance, final DirectBuffer targetElementId) {
+    final ElementInstance flowScope;
+    if (processInstance.getValue().getElementIdBuffer().equals(targetElementId)) {
+      flowScope = processInstance;
+    } else {
+      flowScope = findFlowScopeInChildren(processInstance, targetElementId);
+    }
+    return flowScope;
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java:74]: 🔧 This is just nitpicking, but there's no need for single return in Java code. Early returns are generally favored instead.

```java
if (processInstance.getValue().getElementIdBuffer().equals(targetElementId)) {
  return processInstance;
}

return findFlowScopeInChildren(processInstance, targetElementId);
```

File: engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  public ProcessInstanceCreationCreatedApplier(
+      final MutableProcessState processState,
+      final MutableElementInstanceState elementInstanceState) {
+    this.processState = processState;
+    this.elementInstanceState = elementInstanceState;
+  }
+
+  @Override
+  public void applyState(final long key, final ProcessInstanceCreationRecord value) {
+    if (value.hasStartInstructions()) {
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java:37]: 🔧 nitpick: you could consider an early return here by inverting this if statement.

File: engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          .map(instruction -> process.getElementById(instruction.getElementId()))
+          .filter(element -> element.getElementType().equals(BpmnElementType.PARALLEL_GATEWAY))
+          .map(ExecutableFlowNode.class::cast)
+          .forEach(
+              element -> {
+                final var parentElementId = element.getFlowScope().getId();
+                final ElementInstance flowScope =
+                    findParentFlowScope(processInstance, parentElementId);
+                incrementNumberOfTakenSequenceFlows(element, flowScope);
+              });
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java:53]: 👍 Nice and clean, I love it 😍

File: engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  private void incrementNumberOfTakenSequenceFlows(
+      final ExecutableFlowNode element, final ElementInstance flowScope) {
+    element
+        .getIncoming()
+        .forEach(
+            incoming ->
+                elementInstanceState.incrementNumberOfTakenSequenceFlows(
+                    flowScope.getKey(), element.getId(), incoming.getId()));
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java:105]: 🔧 Please update the PR description, to make it clear that we now have to do this as well.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java [hunk 7]
```diff
@@ -250,26 +250,7 @@ public void takeSequenceFlow(
         context.copy(
             sequenceFlowKey, followUpInstanceRecord, ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN);
 
-    if (canActivateTargetElement(context, target)) {
-      activateElementInstanceInFlowScope(sequenceFlowTaken, target);
-    }
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java:255]: 🔧 Please document the new parallel gateway logic

**Review by korthout — APPROVED** [review]
🚀  Nice work @remcowesterhoud 

🔧 Please make it clear in the PR description what this PR changes.
🔧 Please document the new Parallel Gateway behavior.

💭 This will probably not backport so easily to the older versions, because of the new event applier (those versions don't know about start instructions).

👍 LGTM

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/ParallelGatewayTest.java [hunk 8]
```diff
@@ -349,6 +350,45 @@ public void shouldSplitWithUncontrolledFlow() {
...
+            .onlyCommandRejections()
+            .limit(1)
+            .getFirst();
+
+    assertThat(rejection.getRejectionType()).isEqualTo(RejectionType.INVALID_STATE);
+    assertThat(rejection.getRejectionReason())
+        .isEqualTo(
+            "Expected to be able to activate "
+                + "parallel gateway 'joining', but not all sequence flows have been taken. "
+                + "This could occur when a sequence flow is taken multiple times.");
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/ParallelGatewayTest.java:389]: 🔧 Let's also add to the test that the gateway is activated exactly once.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java (hunk 7 above)
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java:255]: I will do this in a separate PR

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors retry

**Comment by korthout** [comment]
@remcowesterhoud Please add this to the [release notes](https://drive.google.com/drive/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

**Pull request merged**